pkg/storegateway/bucket.go

```diff
@@ -1530,14 +1534,15 @@ func (s *bucketBlockSet) getFor(mint, maxt, maxResolutionMillis int64, blockMatc
 // bucketBlock represents a block that is located in a bucket. It holds intermediate
 // state for the block on local disk.
 type bucketBlock struct {
-	userID     string
-	logger     log.Logger
-	metrics    *BucketStoreMetrics
-	bkt        objstore.BucketReader
-	meta       *metadata.Meta
-	dir        string
-	indexCache indexcache.IndexCache
-	chunkPool  pool.Bytes
+	userID         string
+	logger         log.Logger
+	metrics        *BucketStoreMetrics
+	bkt            objstore.BucketReader
+	meta           *metadata.Meta
+	dir            string
+	indexCache     indexcache.IndexCache
+	chunksPool     *chunksPool
```

Comment:
Does it make sense to have a pool of them per each block?

I can imagine this scenario:
- Read request comes in for tenant 1, we allocate tons of chunks and put them into tenant1/blockX's pool.
- Read request comes in for tenant 2, we allocate tons of chunks and put them into tenant2/blockY's pool.
- Garbage collector is triggered and we flush both pools.
- Repeat again, allocate everything again from scratch.

In this situation there's actually no win in using the pool.

Having a global pool could have the penalty of lock contention if many tenants are allocating chunks simulatenously, but I think that if we reach that point (where we care about the lock contention of allocating too many chunks simultaneously) we'll just OOM probably.

Reply:
Don't have the necessary context to know for how long a block stays in memory before being reclaimed by the GC.

But if what you say is true, wouldn't it apply in the same way to the current [bytes chunk pool](https://github.com/grafana/mimir/blob/main/pkg/storegateway/bucket.go#L1540), which is also shared across all block readers?

Reply:
The blocks stay there virtually forever (unless loaded by another store-gateway or deleted)

And yes, I think this would also apply to the bytes chunk pool. WDYT about this @pracucci ?

Reply:
FYI, just rebased with latests main branch changes and moved the chunks pool to a global scope. 

Pinging @pracucci in case you are able to have a look now.